### PR TITLE
[Telemetry][Kuberentes] Distinguish Kubernetes deployment stacks

### PR DESCRIPTION
**Pull request by DmitriGekhtman**
## Why are these changes needed?

Right now, Ray telemetry indicates the majority of Ray's CPU hour usage comes from Ray running within a Kubernetes cluster. However, we have no data on what method is used to deploy Ray on Kubernetes.

This PR enables Ray telemetry to distinguish between three methods of deploying Ray on Kubernetes:

- KubeRay >= 0.4.0
- Legacy Ray Operator with Ray >= 2.1.0
- All other methods

The strategy is to have the operators inject an env variable into the Ray container's environment.
The variable identifies the deployment method.

This PR also modifies the legacy Ray operator to inject the relevant env variable.
A follow-up KubeRay PR changes the KubeRay operator to do the same thing: https://github.com/ray-project/kuberay/pull/562

## Related issue number

Partly addresses
https://github.com/ray-project/ray/issues/28029

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/_private/usage/usage_lib.py [hunk 1]
```diff
@@ -757,8 +757,17 @@ def get_instance_type(node_config):
     except FileNotFoundError:
         # It's a manually started cluster or k8s cluster
         result = ClusterConfigToReport()
-        if "KUBERNETES_SERVICE_HOST" in os.environ:
-            result.cloud_provider = "kubernetes"
+        # Check if we're on Kubernetes
+        if usage_constant.KUBERNETES_SERVICE_HOST_ENV in os.environ:
+            # Check if we're using KubeRay >= 0.4.0.
+            if usage_constant.KUBERAY_ENV in os.environ:
+                result.cloud_provider = usage_constant.PROVIDER_KUBERAY
```
Comment (jjyao) [inline python/ray/_private/usage/usage_lib.py:764]: For k8s, could we know which cloud provider it's using? Is this something we want to record?

File: python/ray/_private/usage/usage_constants.py [hunk 2]
```diff
@@ -49,3 +49,11 @@
 EXTRA_USAGE_TAG_PREFIX = "extra_usage_tag_"
 
 USAGE_STATS_NAMESPACE = "usage_stats"
+
+KUBERNETES_SERVICE_HOST_ENV = "KUBERNETES_SERVICE_HOST"
+KUBERAY_ENV = "RAY_KUBERAY_IN_USE"
```
Comment (jjyao) [inline python/ray/_private/usage/usage_constants.py:54]: normally usage stats related env vars are started with `RAY_USAGE_STATS_`

Comment (DmitriGekhtman) [reply python/ray/_private/usage/usage_constants.py:54]: Thanks for the pointer! Will modify accordingly.

File: python/ray/_private/usage/usage_lib.py (hunk 1 above)
Comment (DmitriGekhtman) [reply python/ray/_private/usage/usage_lib.py:764]: That would be nice to have, but I think it's fundamentally inaccessible without somehow asking the user to give us that bit of data.
No general way to determine which cloud provider you're on from inside a Ray container in a K8s cluster.

File: python/ray/_private/usage/usage_constants.py (hunk 2 above)
Comment (DmitriGekhtman) [reply python/ray/_private/usage/usage_constants.py]: Fixed.

**Review by jjyao — APPROVED** [review]

**Review by rkooo567 — APPROVED** [review]
Is there any easy way to have e2e test for this?

**Comment by DmitriGekhtman** [comment]
> Is there any easy way to have e2e test for this?

I'm going to verify with the legacy operator _manually_ before merging. (We don't have CI integration tests for that operator.)

For the KubeRay operator, I'll add some e2e logic to the CI after merging the relevant KubeRay PR.

**Comment by DmitriGekhtman** [comment]
Test from within a Ray cluster managed by the legacy Ray operator, with these changes:
```
>>> import ray._private.usage.usage_lib as ray_usage_lib
>>> ray_usage_lib.get_cluster_config_to_report("does_not_exist.yaml").cloud_provider
'legacy_ray_operator'
```
Will let the CI run once more to see if there are any issues before merging.

**Comment by DmitriGekhtman** [comment]
Going to merge before stalling mac tests run, as this pr has no bearing on mac support.

**Pull request merged**